### PR TITLE
chore: removed canUndo , canRedo state, added prevStackState, added isGrouping

### PR DESCRIPTION
**Pull request by cesara**

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -152,3 +152,10 @@ test('past zero group tests, subsequent groups', () => {
   void undoManager.redo();
   expect(modifiedValue).to.be.equal(8);
 });
+
+test('start group twice should throw error', () => {
+  undoManager.startGroup();
```
[the remark this answers is not in the document]
Comment (aboodman) [reply src/index.test.ts:157]: That was my request. I can’t think of a use case for nesting. I checked and eg cocoa does allow nesting but I don’t know what the use case is. Do you?

File: src/index.ts [hunk 2]
```diff
@@ -23,15 +23,21 @@ export type ExecuteUndo = {
...
  * Types of object that can be added to the undoRedo stack. (UndoRedo | ExecuteUndo)
  */
 export type AddOptions = UndoRedo | ExecuteUndo;
 
+/**
+ * Type of function that will be called on an UndoRedoStatckState change.
+ */
+export type OnChangeHandlerType =
+  | ((undoRedoStackState: UndoRedoStackState) => void)
+  | undefined;
```
[the remark this answers is not in the document]
Comment (aboodman) [reply src/index.ts:40]: Cesar, to expand he’s suggesting that at the use site you’d say:

onChange: OnChangeHandler|undefined;

Or maybe just:

onChange?: OnChangeHandler; 

… I can never remember how the ? operator works.

File: src/index.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (arv) [reply src/index.test.ts:157]: Yes. It allows composing commands. Lets assume we have a moveX and a moveY command. Then there is a move command that groups moveX & moveY and then there might be a command called rotateAroundPoint which groups a rotate and a move. Now you have nested command groups.

File: src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (arv) [reply src/index.ts:40]: `T?` is the same as `T | undefined`

**Pull request merged**